### ISSUE TITLE
DWG/DXF Import does not work correctly in newer versions (ex: LTR 3.22.6, 3.24.2, 3.24.3)

### ISSUE THREAD

**Issue by micdeluca** [qgis/QGIS#48637]
### What is the bug or the crash?

I noticed that in newer versions of QGIS (ex: LTR 3.22.6, 3.24.2, 3.24.3) compiled with GDAL/OGR 3.4.2 the DWG/DXF import is not done correctly (see attached examples).
DWG/DWX import still works correctly in old versions with GDAL/OGR 3.4.1 (ex: 3.24.1, 3.22.3).
[TEST_DXF_Import.zip](https://github.com/qgis/QGIS/files/8725058/TEST_DXF_Import.zip)

### Steps to reproduce the issue

Try to import any DWG/DXF file.

### Versions

LTR 3.22.6
3.24.2 
3.24.3

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by rouault** [comment]
can you describe precisely what is wrong in newer versions ? I opened your projects and this wasn't obvious to me.
I'd note tat the GDAL version is unrelated as ths DWG/DXF functionality doesn't use it, and there was no change in GDAL in the DXF driver between 3.4.1 and 3.4.2

**Comment by micdeluca** [comment]
You are right, the projects I attached show no significant differences except the appearance of the polylines, point and inserts entities in projects 3.24.2 and 3.22.6 and the non-assignment of the reference system to the texts, points and inserts entities.
![TEST](https://user-images.githubusercontent.com/65657745/169707797-3a29f20a-1e49-4518-9219-5925674972d6.png)

It is my mistake not to have reopened the projects after creating them.
However, when I created the projects with versions 3.24.2 and 3.22.6, the numbers were not visible, as in the following image.

![TEST_QGIS_3 24 2](https://user-images.githubusercontent.com/65657745/169707852-2a23a164-8056-42c1-b2a6-a721b6cfcefb.png)

The numbers "reappeared" after saving the projects.

**Comment by kadarivan** [comment]
It looks like there is an indexing error in the gpkg loading function.
3.28.6-Firenze - Win10
I made example dwg files with 1-1 elements, and with combinations. The result depends on what elements are in the file:

line:
![QGIS line](https://github.com/qgis/QGIS/assets/18245793/00a8c45f-08d4-49ff-8055-b334a0abc270)
polyline:
![QGIS polyline](https://github.com/qgis/QGIS/assets/18245793/11bb5bbf-3ce5-4f08-a3d4-e245be7acf6b)
text:
![QGIS text](https://github.com/qgis/QGIS/assets/18245793/8028355f-116d-481c-9bc3-8cc5c26bc496)
hatch:
![QGIS hatch](https://github.com/qgis/QGIS/assets/18245793/8ac01e69-51ac-4227-843a-93ef35d40123)
polyline+line:
![QGIS polyline+line](https://github.com/qgis/QGIS/assets/18245793/b231b1cc-8f46-49ff-95b2-352da9e73fef)
polyline+text:
![QGIS polyline+text](https://github.com/qgis/QGIS/assets/18245793/8507c912-a209-43d7-ad9b-6f884835652f)
polyline+line+text:
![QGIS polyline+line+text](https://github.com/qgis/QGIS/assets/18245793/ce563a4e-3e65-451b-b614-8caf6fdf3ba8)
QGIS hatch+line:
![QGIS hatch+line](https://github.com/qgis/QGIS/assets/18245793/fda146ab-5f69-40a8-bf46-de406b9065df)
QGIS hatch+text:
![QGIS hatch+text](https://github.com/qgis/QGIS/assets/18245793/1f9badec-b36a-4367-ac4e-e885a154999b)

Test files (dwg+gpkg+png)
[QGIS3.28 dwg import test.zip](https://github.com/qgis/QGIS/files/11451455/QGIS3.28.dwg.import.test.zip)

**Comment by kadarivan** [comment]
I found some pattern:
The order of layers (objects' type) are fix. If the dwg file contains an object of a certain type, QGIS loads all other layers after that.
- If dwg only contain text, the QGIS load texts and ponts
- if dwg contain hatch, then QGIS load hatches, lines, polylines, texts and points

QGIS always load correctly the first and second layer, but the others are wrong (whether the layer contains an object or not).

**Comment by she-weeds** [comment]
Any update on this issue? It appears that the only fix right now is to import, save project, close and re-open. It does make writing how-to documentation a bit tedious :-) ....

I have observed the same pattern as @kadarivan even in 3.34.1
